### ISSUE TITLE
File upload

### ISSUE THREAD

**Issue by tmulzer** [leihs/leihs#128]
I'm not able to upload files for models ..
I select the local file, it is shown as "Datei wird beim speichern hochgeladen", but when i save and re-open, the file isn't attached to the model.

It did work once (last file-upload was Dec '15).

Here's what's in the log-file:

```
I, [2016-09-06T20:01:53.909659 #13684]  INFO -- : Started POST "/manage/1/models/41/upload/attachment" for 141.64.71.43 at 2016-09-06 20:01:53 +0200
I, [2016-09-06T20:01:53.910642 #13684]  INFO -- : Processing by Manage::ModelsController#upload as */*
I, [2016-09-06T20:01:53.910750 #13684]  INFO -- :   Parameters: {"model"=>{"product"=>"Kleinsteuerung", "version"=>"EASY822-DC-TC", "manufacturer"=>"Eaton", "description"=>"", "technical_detail"=>"", "internal_description"=>"", "hand_over_note"=>"", "category_ids"=>["", "1"]}, "files"=>[#<ActionDispatch::Http::UploadedFile:0x00000000a52e68 @tempfile=#<Tempfile:/tmp/RackMultipart20160906-13684-qxya6h.pdf>, @original_filename="MN04902001Z_DE.pdf", @content_type="application/x-unknown", @headers="Content-Disposition: form-data; name=\"files[]\"; filename=\"MN04902001Z_DE.pdf\"\r\nContent-Type: application/x-unknown\r\n">], "type"=>"attachment", "inventory_pool_id"=>"1", "id"=>"41"}
D, [2016-09-06T20:01:53.911683 #13684] DEBUG -- :   User Load (0.1ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 26 LIMIT 1
D, [2016-09-06T20:01:53.912907 #13684] DEBUG -- :   Language Load (0.1ms)  SELECT  `languages`.* FROM `languages` WHERE `languages`.`id` = 3  ORDER BY `languages`.`name` ASC LIMIT 1
D, [2016-09-06T20:01:53.913567 #13684] DEBUG -- :   InventoryPool Load (0.1ms)  SELECT  `inventory_pools`.* FROM `inventory_pools` WHERE `inventory_pools`.`id` = 1 LIMIT 1
D, [2016-09-06T20:01:53.914553 #13684] DEBUG -- :   AccessRight Load (0.1ms)  SELECT `access_rights`.* FROM `access_rights` WHERE `access_rights`.`user_id` = 26 AND `access_rights`.`inventory_pool_id` = 1 AND `access_rights`.`deleted_at` IS NULL
D, [2016-09-06T20:01:53.915521 #13684] DEBUG -- :   Model Load (0.1ms)  SELECT  `models`.* FROM `models` WHERE `models`.`id` = 41  ORDER BY product ASC LIMIT 1 OFFSET 0
I, [2016-09-06T20:01:53.947256 #13684]  INFO -- : Command :: file -b --mime '/tmp/6d2d7e5c0793c57547eddd2db2bc5ecb20160906-13684-15u7zmp.pdf'
D, [2016-09-06T20:01:53.959829 #13684] DEBUG -- :    (0.1ms)  BEGIN
I, [2016-09-06T20:01:53.968779 #13684]  INFO -- : Command :: file -b --mime '/tmp/6d2d7e5c0793c57547eddd2db2bc5ecb20160906-13684-rze8e7.pdf'
D, [2016-09-06T20:01:53.981093 #13684] DEBUG -- :    (0.1ms)  ROLLBACK
I, [2016-09-06T20:01:53.981751 #13684]  INFO -- : Completed 200 OK in 71ms (ActiveRecord: 0.7ms)

```

Any idea?

Kind regards,
Tasso

**Comment by tmulzer** [comment]
Permissions of the folder .../public/attachments/... should be okay .. I tried it with chmod -R 777 attachments .. didn't help.

**Comment by tmulzer** [comment]
One more thing .. the file-upload of the model-picture did work.

**Comment by nimaai** [comment]
Currently only upload of the following file extensions are supported: `png`, `gif`, `jpeg`, `pdf`. I don't know what the motivation is behind it as we don't have test scenarios for it. Do you @psy-q ?

Either way, we should return an error if the validation of the content type fails and not just silently ignore the attachment like it is done now.

**Comment by tmulzer** [comment]
Hmmm .. I tried to upload pdf files - so it's not the extension.
The System seems to invoke the command `file -b --mime <filename>`, which returns:
`application/pdf; charset=binary` for the uploaded file.

**Comment by nimaai** [comment]
Yes, you are right, I overlooked that. 

I tried to upload a PDF and it worked on my machine and also on our test server. 

Another validation we are having is in regards to file size (upto 100MB). What's the file size of the file you are trying to upload?

**Comment by tmulzer** [comment]
The file has a size of about 5.5 MByte - should not collide with the file size limitation.

It's transferred to the servers /tmp/ folder .. and I can find it there three times .. the filenames are:
`6d2d7e5c0793c57547eddd2db2bc5ecb20160909-6142-10xtolf.pdf
6d2d7e5c0793c57547eddd2db2bc5ecb20160909-6142-ujjj7x.pdf
RackMultipart20160909-6142-1xls35o.pdf`

all of them with permissions 600 and the leihs-user and -group as owner.

When I login to the server later, the files are gone - don't know, if it's ubuntu or leihs, who removes the files.

**Comment by nimaai** [comment]
Probably ubuntu. The system `/tmp` folder also gets cleared on every reboot as far as I know.

As indicated by the folder name itself this location is temporary and is used by leihs (Rack) to process the uploaded file. The final destination of the file is under `<leihs_path>/public/attachments`. Can you please check if you can find your file there? 

But first of all you should check if the model has any attachments on DB/Rails model level. You can go in the rails console (`bundle exec rails console`) and find it out: `Model.find(<model_id>).attachments` and post it here.

If there is such an attachment then we can use the ID of it to try to find the file itself under `public/attachments`, as the attachment id's are used to build the subfolder structure there.

**Comment by tmulzer** [comment]
Hmmm .. can't start the console and get following error message:
`/usr/local/rvm/gems/ruby-2.1.5/gems/mysql2-0.3.20/lib/mysql2/client.rb:70:in `connect': Access denied for user 'root'@'localhost' (using password: NO) (Mysql2::Error)
[...]`

I'm sorry, but I'm no ruby-expert and don't know how to tell it the correct mysql-credentials...

**Comment by nimaai** [comment]
If you start `RAILS_ENV=production bundle exec rails console` from your Rails app folder then the database connection credentials are taken from `./config/database.yml`.

**Comment by tmulzer** [comment]
Here is the result of the console-command:
`2.1.5 :002 > Model.find(41).attachments`
`Model Load (0.1ms)  SELECT  'models'.* FROM 'models' WHERE 'models'.'id' = 41 LIMIT 1`
`Attachment Load (0.1ms)  SELECT 'attachments'.* FROM 'attachments' WHERE 'attachments'.'model_id' = 41`
` => #<ActiveRecord::Associations::CollectionProxy []>`

**Comment by nimaai** [comment]
I think it's being rolled back really because of the content_type. I analyzed your log file again and now I see there `@content_type="application/x-unknown"`. Can you please try to upload some other PDF file?

**Comment by psy-q** [comment]
I'm late to the party, but I remember some issue where the **client** specified the MIME type, and if the MIME type was wrong on the client (either in file metadata or as determined by the client browser), it would also be transmitted wrongly to the server.

This was an issue for example on Macs that had downloaded PDFs with MIME type `application/octet-stream` from broken web applications. For some reason the Macs would store this wrong MIME type with the PDF file and present that to web servers when uploading the same file. Yet locally on the file system, they would use the .pdf file extension instead to open in the correct program.

This is a mess that I believe started when Adobe's PDF reader plugin decided to simply open any PDF presented to it, ignoring the header:

```
Content-Disposition: attachment;
```

That would normally be used to prompt the client to download a file instead of displaying it. That way, the Content-type header that the web app delivers could still correctly be `application/pdf`.

Not that this story helps much, but I believe the mess might happen on the client. And thanks, Adobe.

**Comment by tmulzer** [comment]
Okay - some time went by - I fixed it now. It's really a firefox-/acrobat-induced problem.

Following site helped:
https://bugzilla.mozilla.org/show_bug.cgi?id=373621

I removed all parts of Adobe Acrobat from the "mimeTypes.rdf" in the directory "%APPDATA%\Mozilla\" - content type of the upload is application/pdf and it works again.
Different solution was using IExplorer or trying other browsers - but that wasn't my preferred solution.

Kind regards,
Tasso
